### DISCUSSION TITLE
How should tabs work?

### DISCUSSION THREAD

**Discussion by WaldemarH** [OpenSalamander/salamander#60] [Ideas]
Hi guys,

as I'm slowly getting the feel for the OS code I started thinking how to implement tabs.
Now as OS already has 'switch panel' functionality we really already have all that we need... well we'll see.

My idea is that I would put a 'Tabs' submenu right on top of the disk dropdown menu -> check image.
Where submenu would contain:
1. 'Add tab...' menu item -> a dialog would open where you could name the tab
2. 'Remove tab...' menu item
3. separator
4. Tabs

![image](https://github.com/user-attachments/assets/37df592b-5d92-4207-8cf9-a830250b0d81)

Now every tab would stay in memory and would just be hidden away, so when you would select a tab on a left/right panel current one would hide and the selected one would show.

So this would give us the first version of tabs. In time we would pimp it, but for the intro it should be good enough.

What do you think?

BrW

**Comment by melloware** [comment]
Interesting so are you saying Tabs would swap panels out?

**Reply by WaldemarH** [reply]
That would be the most useful thing.. so all the selections,.. would stay as they are.
The left and right panels would just be hidden away and 'a new ones created' or 'a hidden ones shown'.

All in all the vivaldi's tab managements is quite useful especially if you are working on multiple projects.

**Comment by WaldemarH** [comment]
Was thinking a little more and I'll probably just recreate the Vivaldi's tab management... I really like the grouping of the tabs. The only difference will be that a tab selection would swap both panels.

![image](https://github.com/user-attachments/assets/8afed0d0-aada-406b-b54e-b0ef880f9e17)

Now as the whole upper part of OS is a rebar we would add a full width row window where tabs would be defined.

In time we could then add some config options on how tabs behave... like the selected tab is wider and different color,...

BrW

**Reply by melloware** [reply]
slick

**Reply by WaldemarH** [reply]
Yeah, was also thinking that later we can add even more flexibility with a separate tab management for every panel. So we would have a "global" tabs which would switch both panels and then every separate tab would also contain "sub-tabs". For example a lot of times I have a file that I need to copy to two or more places and I could select a Project and have this "global tab" that would have on the left side the source folder and on the right multiple tabs with all the destinations".
Now one thing we'll have to be careful to not use to much screen view... I still need to think a littl more on this, but probably the "sub-tabs" would be integrated into the 'disk dropdown menu' as shown on the first image.

**Comment by tukanos** [comment]
I think tab grouping and having a tab management like Vivaldi is marvelous idea.  It would be also nice to be able to backup such tabs.

As for visuals, it is indeed important to take up as little space as possible.   What about this idea?  You could have a configuration item where you would "see" the sub-tabs only if you hoover over them with mouse.  If you leave the area it could retract back.

**Reply by WaldemarH** [reply]
Lets roll on this... the area left of the path could be used for "local panel only" tabs... a rough drawing...
1. Clicking on it would give you a dropdown menu with a name and path (maybe we could add a config what to show a name only, path only, name and path together), so we could have A LOT of local tabs per panel if needed.
2. Right click would give you the classical add, modify,... context menu.

![image](https://github.com/user-attachments/assets/0cf8ca7a-c8ef-4eed-a14d-7b2001fa24f5)

Also right click on a path would include an option to "Add to tabs" where a dialog would open where you could define a name.

Now all the global and local tabs would be stored and in settings you would have management where you could change the path/delete/reorder/...

Also all the local tabs would be directly linked to the selected global tab. So every global tab selection would have its own local "tabs".


So we have multiple layers of paths available to us:
1. global hot paths
2. history (will also get a redesign)
2. tabs

**Comment by tukanos** [comment]
You could have such small space for the tabs.  Second option would be to have a whole line of tabs instead (like browser does)of e.g. the quick icons (right below the file menu).  Where you could have the arrow down (that would be local tabs?) for any additional tabs.  

It could be moved like like the other layers.

**Reply by WaldemarH** [reply]
The good side of everything is that we can have both.. we can just add an option what kind of tabs you want. :)
As lots of things in AS/OS there is a checkbox to enable(show)/disable(hide).

**Reply by WaldemarH** [reply]
THe only thing that we really need to define is what we really want.

**Reply by tukanos** [reply]
@WaldemarH Yup, that is great.  I think the most important part is that it has to follow "natural" workflow.  Vivaldi's tabs are great, for example, moving the single tab into a group is a killer feature that I use often.  Now the question is, how many people are actually clicking on the action icons in Salamander.  I have never ever used that feature.  I have always used shortcuts so for me that space is great way to use it.  

In my eyes here is the space (one is the part where you placed your tab):
![salamander_tabs](https://github.com/user-attachments/assets/290eaa7d-39b1-4fe1-8b0d-eb00107031b3)

Note: that is not my working Salamander, that is one on some VM.

**Reply by melloware** [reply]
I use the icons ALL the time especially the Zip icon where I highlight some files nd click the Compress Icon to zip them.

**Reply by tukanos** [reply]
@melloware I see.  I'm using the keyboard shortcut for that.  One time I had it on quick start where I had all different kinds of compression(s).

**Reply by melloware** [reply]
And the toolbar can be hidden anyway with an option already right?

**Reply by WaldemarH** [reply]
For now I was thinking like this:

![image](https://github.com/user-attachments/assets/96f51b4f-9b78-434c-9f55-2a95bcabc63e)

**Reply by melloware** [reply]
i like it

**Reply by tukanos** [reply]
@melloware Yes, it can be hidden with a configuration.

@WaldemarH Yes, that looks great!  That would enable everyone to have it as they want.

**Reply by melloware** [reply]
@WaldemarH when do @tukanos and I get your BETA? 😄

**Comment by KRtekTM** [comment]
Hello there,

You can try and test the Tabbed panels feature which I'm currently implementing using OpenAI Codex: https://github.com/OpenSalamander/salamander/pull/168

I found this topic just now, so my work is not inspired by the discussion here.
Just a short summary after I quickly read it:
- I don't like any dropdown menu as tab selection because it slows the hand performing mouse movement and therefore your work (ugly UI is often more usable than clean one)
- The tab toolbar can't be move as it's related to panels, so it's directly above them.
- Tabs are only per-panel. If you need "global tabs", you need to run second Salamander instance. No plan to change it since the tabbed panels implementation is already huge and modifies a big part of original source code.
